nebula_rust/src/graph_client/connection_pool.rs

```diff
@@ -4,4 +4,143 @@
  * attached with Common Clause Condition 1.0, found in the LICENSES directory.
  */
 
-mod graph_client {};
+use crate::graph_client::connection::Connection;
+use crate::graph_client::pool_config::PoolConfig;
+use crate::graph_client::session::Session;
+
+/// The pool of connection to server, it's MT-safe to access.
+pub struct ConnectionPool {
+    /// The connections
+    /// The interior mutable to enable could get multiple sessions in one scope
+    conns: std::sync::Mutex<std::cell::RefCell<std::collections::LinkedList<Connection>>>,
+    /// It should be immutable
+    config: PoolConfig,
+    /// Address cursor
+    cursor: std::cell::RefCell<std::sync::atomic::AtomicUsize>,
+    /// The total count of connections, contains which hold by session
+    conns_count: std::cell::RefCell<std::sync::atomic::AtomicUsize>,
+}
+
+impl ConnectionPool {
+    /// Construct pool by the configuration
+    pub async fn new(conf: &PoolConfig) -> Self {
+        let conns = std::collections::LinkedList::<Connection>::new();
+        let pool = ConnectionPool {
+            conns: std::sync::Mutex::new(std::cell::RefCell::new(conns)),
+            config: conf.clone(),
+            cursor: std::cell::RefCell::new(std::sync::atomic::AtomicUsize::new(0)),
+            conns_count: std::cell::RefCell::new(std::sync::atomic::AtomicUsize::new(0)),
+        };
+        assert!(pool.config.min_connection_pool_size <= pool.config.max_connection_pool_size);
+        pool.new_connection(pool.config.min_connection_pool_size)
+            .await;
+        pool
+    }
+
+    /// Get a session authenticated by username and password
+    /// retry_connect means keep the connection available if true
+    pub async fn get_session(
+        &self,
+        username: &str,
+        password: &str,
+        retry_connect: bool,
+    ) -> std::result::Result<Session<'_>, common::types::ErrorCode> {
+        if self.conns.lock().unwrap().borrow_mut().is_empty() {
+            self.new_connection(1).await;
+        }
+        let conn = self.conns.lock().unwrap().borrow_mut().pop_back();
```

Comment:
the connection pop from the list, and if the user does no return it, the pool couldn't close it. You need to make the pool can manage it when the use get it from the pool.

Reply:
you need to make the connection is ok, here you can get a bad connection

Reply:
The pool can't access the connection owned by session. It's designed to avoid some data race

Reply:
The keep-alive will add in later pr.